muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/guider/GuiderRESTService.java

```diff
@@ -1045,4 +1086,175 @@ else if (value instanceof Double) {
+
+    if (loggedUserRole == null) {
+      return Response.status(Status.BAD_REQUEST).build();
+    }
+
+    if (!userSchoolDataController.amICounselor(studentIdentifier) && !loggedUserRole.equals(EnvironmentRoleArchetype.ADMINISTRATOR)) {
+      return Response.status(Status.UNAUTHORIZED).build();
+    }
+
+
+    WorkspaceEntity workspaceEntity = workspaceController.findWorkspaceEntityById(workspaceEntityId);
+    if (workspaceEntity == null) {
+      return Response.status(Status.BAD_REQUEST).build();
+    }
+
+    User student = userController.findUserByDataSourceAndIdentifier(studentIdentifier.getDataSource(), studentIdentifier.getIdentifier());
+    
+    if (student == null) {
+      return Response.status(Status.NOT_FOUND).build();
+    }
+
+    Workspace workspace = workspaceController.findWorkspace(workspaceEntity);
+
+    SchoolDataIdentifier workspaceIdentifier = new SchoolDataIdentifier(workspace.getIdentifier(), workspace.getSchoolDataSource());
+
+    WorkspaceUserEntity workspaceUserEntity = workspaceUserEntityController.findWorkspaceUserEntityByWorkspaceAndUserIdentifierIncludeArchived(workspaceEntity, studentIdentifier);
+    if (workspaceUserEntity != null && workspaceUserEntity.getArchived() == Boolean.TRUE) {
+      workspaceUserEntityController.unarchiveWorkspaceUserEntity(workspaceUserEntity);
+    }
+    if (workspaceUserEntity != null && workspaceUserEntity.getActive() == Boolean.FALSE) {
+      workspaceUserEntityController.updateActive(workspaceUserEntity, Boolean.TRUE);
+      userIndexer.indexUser(workspaceUserEntity.getUserSchoolDataIdentifier().getUserEntity());
+    }
+
+    fi.otavanopisto.muikku.schooldata.entity.WorkspaceUser workspaceUser = workspaceController.findWorkspaceUserByWorkspaceAndUser(workspaceIdentifier, studentEntity.defaultSchoolDataIdentifier());
+    if (workspaceUser == null) {
+      workspaceUser = workspaceController.createWorkspaceUser(workspace, student, WorkspaceRoleArchetype.STUDENT);
+      waitForWorkspaceUserEntity(workspaceEntity, studentIdentifier);
+    }
+    else {
+      workspaceController.updateWorkspaceStudentActivity(workspaceUser, true);
+    }
+
+    List<WorkspaceUserEntity> workspaceTeachers = workspaceUserEntityController.listActiveWorkspaceStaffMembers(workspaceEntity);
+    List<UserEntity> recipients = new ArrayList<UserEntity>();
+
+    String workspaceName = workspace.getName();
+    if (!StringUtils.isBlank(workspace.getNameExtension())) {
+      workspaceName += String.format(" (%s)", workspace.getNameExtension()); 
+    }
+
+    String userName = student.getNickName() == null
+      ? student.getDisplayName()
+      : String.format("%s \"%s\" %s (%s)", student.getFirstName(), student.getNickName(), student.getLastName(), student.getStudyProgrammeName());
+
+    User loggedUser = userController.findUserByUserEntityDefaults(sessionController.getLoggedUserEntity());
```

Comment:
No need to fetch logged user all the way from Pyramus just to determine their name. Use `userEntityController.getName` instead.